tmt/result.py

```diff
@@ -317,49 +337,77 @@ def from_test_invocation(
             ids=ids,
             log=log or [],
             guest=ResultGuestData.from_test_invocation(invocation=invocation),
-            data_path=invocation.relative_test_data_path)
+            data_path=invocation.relative_test_data_path,
+            check=invocation.check_results or [])
 
-        return _result.interpret_result(invocation.test.result)
+        interpret_checks = {check.how: check.result for check in invocation.test.check}
 
-    def interpret_result(self, interpret: ResultInterpret) -> 'Result':
+        return _result.interpret_result(invocation.test.result, interpret_checks)
+
+    def interpret_result(
+            self,
+            interpret: ResultInterpret,
+            interpret_checks: dict[str, CheckResultInterpret]
+            ) -> 'Result':
         """
         Interpret result according to a given interpretation instruction.
 
         Inspect and possibly modify :py:attr:`result` and :py:attr:`note`
         attributes, following the ``interpret`` value.
 
         :param interpret: how to interpret current result.
+        :param interpret_checks: mapping of check:how and it's result interpret
         :returns: :py:class:`Result` instance containing the updated result.
         """
 
-        if interpret in (ResultInterpret.RESPECT, ResultInterpret.CUSTOM):
+        if interpret not in ResultInterpret:
+            raise tmt.utils.SpecificationError(
+                f"Invalid result '{interpret.value}' in test '{self.name}'."
+                )
+
+        if interpret == ResultInterpret.CUSTOM:
             return self
 
-        # Extend existing note or set a new one
-        if self.note:
-            self.note += f', original result: {self.result.value}'
+        original_result = self.result
+
+        # Interpret check results
+        self.check = [
+            check_result.interpret_check_result(CheckResultInterpret(interpret_checks[check_result.name]))
```

Comment:
```suggestion
            check_result.interpret_check_result(interpret_checks[check_result.name])
```

The `interpret_checks` values are already `CheckResultInterpret` objects:

```python
(Pdb) pp interpret_checks
{'avc': <CheckResultInterpret.XFAIL: 'xfail'>,
 'dmesg': <CheckResultInterpret.RESPECT: 'respect'>}

(Pdb) interpret_checks[self.check[0].name]
<CheckResultInterpret.RESPECT: 'respect'>
```

So we don't need to instantiate them again, right?

Reply:
Let's explicitly mention this in the docstring as well. It was a bit confusing to understand.

Reply:
> The interpret_checks values are already CheckResultInterpret objects:

Correct, it's exactly as declared in the method signature.

> Let's explicitly mention this in the docstring as well. It was a bit confusing to understand.

I wouldn't :( I know it sounds tempting, but I feel it a sign of resignation :) The variable is a simple mapping, IDEs, and Sphinx should be able to extract and display this information from the type of the parameter, and, more importantly, at least to me, there should be not even the slightest shadow of a doubt about what type of values in `interpret_checks` would be.

It's the plague of Python projects, this endless second-guessing. That's why it's so important to establish such a type once and for all, on boundaries between tmt and the outside world, and to fight with sword and flame against any ambiguity, to never ever allow the "check result interpretation" values to be anything else but that enum. There must be no question about it on our minds, as soon as you see a name like `interpret_checks`, never should the question "hmm, what are the values, strings maybe?" appear. I for one am not very fond of explicitly pointing out a singular type in docstrings, because it always feels like resignation on this core principle and return of doubt and second-guessing :(

Reply:
Ah! Right you are, in your fight, of course! For some reason I had my blind spot on exactly this line:

```python
            interpret_checks: dict[str, CheckResultInterpret]
```

Don't know why, don't know how it happened. Now I see it, and the world is in order again :-)